### PR TITLE
soc: esp32: add reboot call

### PR DESCRIPTION
**Pull request by sylvioalves**
Add missing ESP32 reset function

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

### CODE REVIEW

**Comment by sylvioalves** [comment]
It depends on #30424

File: soc/xtensa/esp32/soc.c [hunk 1]
```diff
@@ -16,6 +16,28 @@
 #include <toolchain/gcc.h>
 #include <zephyr/types.h>
 
+#include <string.h>
+#include "esp_system.h"
+#include "esp_private/system_internal.h"
+#include "esp_attr.h"
+#include "esp_efuse.h"
+#include "esp_wifi.h"
```
Comment (mniestroj) [inline soc/xtensa/esp32/soc.c:24]: This list of includes seems too big. `string.h` is includes second time, `esp_efuse.h` and `esp_wifi.h` do not seem to be required. Please use minimal list of includes.

File: soc/xtensa/esp32/soc.c [hunk 2]
```diff
@@ -36,28 +58,29 @@ void __attribute__((section(".iram1"))) __start(void)
 	__asm__ __volatile__ (
 		"wsr %0, vecbase"
 		:
-		: "r"(&_init_start));
+		: "r" (&_init_start));
```
Comment (mniestroj) [inline soc/xtensa/esp32/soc.c:61]: Those whitespace changes do not affect "reboot" functionality, so they should not be part of the commit.

File: soc/xtensa/esp32/soc.c [hunk 3]
```diff
@@ -84,3 +107,75 @@ int arch_printk_char_out(int c)
...
+	const uint32_t core_id = cpu_hal_get_core_id();
+	const uint32_t other_core_id = (core_id == 0) ? 1 : 0;
+
+	soc_ll_reset_core(other_core_id);
+	soc_ll_stall_core(other_core_id);
+
+	/* Flush any data left in UART FIFOs */
+	esp32_rom_uart_tx_wait_idle(0);
+	esp32_rom_uart_tx_wait_idle(1);
+	esp32_rom_uart_tx_wait_idle(2);
```
Comment (mniestroj) [inline soc/xtensa/esp32/soc.c:130]: What is the reason to flush UART before doing reset?

File: soc/xtensa/esp32/soc.c [hunk 4]
```diff
@@ -84,3 +107,75 @@ int arch_printk_char_out(int c)
...
+		soc_ll_reset_core(1);
+		soc_ll_reset_core(0);
+	} else {
+		/* Running on APP CPU: need to reset PRO CPU and unstall it, */
+		/* then reset APP CPU */
+		soc_ll_reset_core(0);
+		soc_ll_stall_core(0);
+		soc_ll_reset_core(1);
+	}
+	while (true) {
```
Comment (mniestroj) [inline soc/xtensa/esp32/soc.c:178]: Please add new line between if and while.

File: soc/xtensa/esp32/soc.c (hunk 2 above)
Comment (sylvioalves) [reply soc/xtensa/esp32/soc.c:61]: We have fixed a bunch of code style in another commit. This diff will then be bypassed.

File: soc/xtensa/esp32/soc.c (hunk 3 above)
Comment (sylvioalves) [reply soc/xtensa/esp32/soc.c:130]: This is part of esp-idf implementation. Perhaps a way to avoid uart output trash after reset.

**Review by dcpleung — APPROVED** [review]

**Review by mniestroj — APPROVED** [review]

**Review by nashif — APPROVED** [review]
